docker/Dockerfile.silverstripe

```
@@ -143,7 +143,9 @@ COPY public /var/www/public
 COPY app /var/www/app
 
 # Install application
-RUN composer install 
+RUN composer install --no-scripts
+RUN composer vendor-expose
+RUN npm install --dev
```

Again, `--dev` is deprecated, and causes the installation of only `devDependencies` which is not typically what we want.

Also

```suggestion
RUN npm ci --dev
```

---

🐘 you can optimise things by copy `package.json` & `package-lock.json` and doing install, before copying the rest of the application, since the whole app shouldn;'t be needed for installing.

---

@G-Rath what do you mean that it shouldn't be needed for installing?

Also, thanks for pointing these out: I realised that there's no process where the package-lock.json is being created after cloning the project D:

---

> what do you mean that it shouldn't be needed for installing?

The nature of Node means that to install your dependencies you need at least one thing: the `package.json`; ideally you'll have two things: `package.json` + the lockfile (`package-lock.json`, `yarn.lock`, or `pnpm-lock.yml`).

This means that you can do the `npm install` step before copying the rest of the application files, which in docker land means that changes to application files that are not the two files mentioned above won't invalid the `npm install` layer which typically is one of the more time consuming commands.

This in theory applies to `composer install` as well, except that `composer` packages can be copied into folders and in general result in changes that lie outside of the `vendor` folder.

You still shouldn't require all the application *files* for a `composer install`, but you may need the folder structure which might be annoying to replicate manually in the dockerfile (but still worth it if it's easy, since its another long running command).

> I realised that there's no process where the package-lock.json is being created after cloning the project D:

[**You should always commit your lockfile**](https://classic.yarnpkg.com/blog/2016/11/24/lockfiles-for-all/). This ensures that we know the exact versions of packages the project will work on, and so makes it easy to determine when dependency updates break something.

This also gives us a baseline for auditing, and lets us use `npm ci` for fast installs (as it *requires* a lockfile, so these commands will fail if one doesn't already exist).

---

I've addressed the lock files issue in another branch